File: src/main/scala/com/xmlcalabash/util/MediaType.scala
package com.xmlcalabash.util

import java.net.{URI, URLConnection}
import com.xmlcalabash.exceptions.XProcException

import java.util.regex.Pattern
import scala.collection.mutable.ListBuffer

// This isn't trying very hard to be strict about the rules
// N.B. This class accepts "*" for type and subtype because it's used for matching

object MediaType {
  val OCTET_STREAM = new MediaType("application", "octet-stream")
  val TEXT = new MediaType("text", "plain")
  val XML  = new MediaType("application", "xml")
  val JSON = new MediaType("application", "json")
  val YAML = new MediaType("application", "vnd.yaml")
  val HTML = new MediaType("text", "html")
  val XHTML = new MediaType("application", "xhtml+xml")
  val ZIP = new MediaType("application", "zip")
  val MULTIPART = new MediaType("multipart", "*")
  val MULTIPART_MIXED = new MediaType("multipart", "mixed")

  val MATCH_XML: Array[MediaType] = Array(
    MediaType.parse("application/xml"),
    MediaType.parse("text/xml"),
    MediaType.parse("*/*+xml"),
    MediaType.parse("-application/xhtml+xml"))

  val MATCH_HTML: Array[MediaType] = Array(
    MediaType.parse("text/html"),
    MediaType.parse("application/xhtml+xml")
  )

  val MATCH_TEXT: Array[MediaType] = Array(
    MediaType.parse("text/*"),
    MediaType.parse("application/relax-ng-compact-syntax"),
    MediaType.parse("application/xquery"),
    MediaType.parse("application/javascript"),
    MediaType.parse("-text/html"),
    MediaType.parse("-text/xml")
  )

  val MATCH_JSON: Array[MediaType] = Array(
    MediaType.parse("application/json"),
    MediaType.parse("*/*+json")
  )

  val MATCH_YAML: Array[MediaType] = Array(
    MediaType.parse("application/vnd.yaml")
  )

  val MATCH_ANY: Array[MediaType] = Array(
    MediaType.parse("*/*")
  )

  def parse(mtype: Option[String]): Option[MediaType] = {
    if (mtype.isDefined) {
      Some(parse(mtype.get))
    } else {
      None
    }
  }

  def parse(mtype: String): MediaType = {
    parse(mtype, None)
  }

  def parse(mtype: String, forceEncoding: Option[String]): MediaType = {
    // [-]type/subtype; name1=val1; name2=val2
    var pos = mtype.indexOf("/")
    var mediaType = if (pos < 0) {
      mtype.trim
    } else {
      mtype.substring(0, pos).trim
    }

    var rest = if (pos < 0) {
      ""
    } else {
      mtype.substring(pos + 1).trim
    }

    var inclusive = true
    if (mediaType.startsWith("-")) {
      inclusive = false
      mediaType = mediaType.substring(1)
    }

    // This is a bit convoluted because of the way 'rest' is reused.
    // There was a bug and this was the easiest fix. #hack

    var params = ""
    pos = rest.indexOf(";")
    if (pos >= 0) {
      params = rest.substring(pos+1)
      rest = rest.substring(0, pos).trim
    }

    var mediaSubtype = rest
    var suffix = Option.empty[String]

    if (mediaSubtype.contains("+")) {
      pos = mediaSubtype.indexOf("+")
      suffix = Some(mediaSubtype.substring(pos+1).trim)
      mediaSubtype = mediaSubtype.substring(0, pos).trim
    }

    if (forceEncoding.isEmpty && params == "") {
      new MediaType(mediaType, mediaSubtype, suffix, inclusive, None)
    } else {
      rest = params
      pos = rest.indexOf(";")
      val plist = ListBuffer.empty[String]
      if (forceEncoding.isDefined) {
        plist += "charset=" + forceEncoding.get
      }

      while (pos >= 0) {
        val param = rest.substring(0, pos).trim
        rest = rest.substring(pos+1)
        if (param != "") {
          if (forceEncoding.isEmpty || !param.startsWith("charset=")) {
            plist.append(param)
          }
        }
        pos = rest.indexOf(";")
      }
      if (rest.trim != "") {
        rest = rest.trim
        if (forceEncoding.isEmpty || !rest.startsWith("charset=")) {
          plist.append(rest)
        }
      }
      new MediaType(mediaType, mediaSubtype, suffix, inclusive, Some(plist.toArray))
    }
  }

  def parseList(ctypes: String): ListBuffer[MediaType] = {
    val contentTypes = ListBuffer.empty[MediaType]
    for (ctype <- ctypes.split("\\s+")) {
      ctype match {
        case "xml"  => contentTypes ++= MATCH_XML
        case "html" => contentTypes ++= MATCH_HTML
        case "text" => contentTypes ++= MATCH_TEXT
        case "json" => contentTypes ++= MATCH_JSON
        case "any"  => contentTypes ++= MATCH_ANY
        case _      =>
          if (ctype.indexOf("/") <= 0) {
            throw XProcException.xsUnrecognizedContentTypeShortcut(ctype, None)
          }
          contentTypes += MediaType.parse(ctype)
      }
    }
    contentTypes
  }
}

class MediaType(val mediaType: String, val mediaSubtype: String, val suffix: Option[String], val inclusive: Boolean, val param: Option[Array[String]]) {
  def this(mediaType: String, mediaSubtype: String) = {
    this(mediaType, mediaSubtype, None, true, None)
  }

  def this(mediaType: String, mediaSubtype: String, suffix: String) = {
    this(mediaType, mediaSubtype, Some(suffix), true, None)
  }

  def parameters: List[String] = {
    if (param.isEmpty) {
      List.empty[String]
    } else {
      val names = ListBuffer.empty[String]
      for (aparam <- param.get) {
        val pos = aparam.indexOf("=")
        if (pos > 0) {
          names += aparam.substring(0, pos)
        }
      }
      names.toList
    }
  }

  def discardParams(): MediaType = {
    if (param.isEmpty) {
      return this
    }
    discardParams(parameters)
  }

  def discardParams(exclude: List[String]): MediaType = {
    if (param.isEmpty) {
      return this
    }

    var found = false
    val newParam = ListBuffer.empty[String]
    for (aparam <- param.get) {
      var pmatch = false
      for (excl <- exclude) {
        val pstr = s"$excl="
        pmatch = pmatch || aparam.startsWith(pstr)
      }
      found = found || pmatch
      if (!pmatch) {
        newParam += aparam
      }
    }

    if (found) {
      if (newParam.isEmpty) {
        new MediaType(mediaType, mediaSubtype, suffix, inclusive, None)
      } else {
        new MediaType(mediaType, mediaSubtype, suffix, inclusive, Some(newParam.toArray[String]))
      }
    } else {
      this
    }
  }

  def addParam(name: String, value: String): MediaType = {
    val newParam = ListBuffer.empty[String]
    if (param.isDefined) {
      val pstr = s"$name="
      for (aparam <- param.get) {
        if (!aparam.startsWith(pstr)) {
          newParam += aparam
        }
      }
    }
    newParam += s"$name=$value"
    new MediaType(mediaType, mediaSubtype, suffix, inclusive, Some(newParam.toArray[String]))
  }

  def classification: MediaType = {
    if (xmlContentType) {
      MediaType.XML
    } else if (htmlContentType) {
      MediaType.HTML
    } else if (textContentType) {
      MediaType.TEXT
    } else if (jsonContentType) {
      MediaType.JSON
    } else if (yamlContentType) {
      MediaType.YAML
    } else {
      MediaType.OCTET_STREAM
    }
  }

  def textContentType: Boolean = {
    val mtype = matchingMediaType(MediaType.MATCH_TEXT)
    mtype.isDefined && mtype.get.inclusive
  }

  def xmlContentType: Boolean = {
    val mtype = matchingMediaType(MediaType.MATCH_XML)
    mtype.isDefined && mtype.get.inclusive
  }

  def jsonContentType: Boolean = {
    val mtype = matchingMediaType(MediaType.MATCH_JSON)
    mtype.isDefined && mtype.get.inclusive
  }

  def yamlContentType: Boolean = {
    val mtype = matchingMediaType(MediaType.MATCH_YAML)
    mtype.isDefined && mtype.get.inclusive
  }

  def htmlContentType: Boolean = {
    val mtype = matchingMediaType(MediaType.MATCH_HTML)
    mtype.isDefined && mtype.get.inclusive
  }

  def anyContentType: Boolean = true

  def markupContentType: Boolean = {
    xmlContentType || htmlContentType
  }

  def matchingMediaType(mtypes: Array[MediaType]): Option[MediaType] = {
    var matching = Option.empty[MediaType]
    for (mtype <- mtypes) {
      //println(s"$this $mtype ${matches(mtype)}")
      if (matches(mtype)) {
        matching = Some(mtype)
      }
    }
    matching
  }

  def matches(mtype: MediaType): Boolean = {
    if (mtype.mediaType == "application" && mtype.mediaSubtype == "octet-stream") {
      return true
    }

    var mmatch = mediaType == mtype.mediaType || mtype.mediaType == "*"
    mmatch = mmatch && (mediaSubtype == mtype.mediaSubtype || mtype.mediaSubtype == "*")
    if (suffix.isDefined && mtype.suffix.isDefined) {
      mmatch = mmatch && suffix.get == mtype.suffix.get
    }

    // application/xml should match */*
    // but text/plain shouldn't match */*+xml

    // This special rule seems necessary but I can't really justify it
    if (mmatch && mtype.mediaType == "*" && mtype.mediaSubtype == "*" && mtype.suffix.isDefined) {
      mmatch = mmatch && suffix.isDefined
    }

    mmatch
  }

  def allowed(types: List[MediaType]): Boolean = {
    // This media type is allowed if it's allowed by at least one member of types
    // and is not forbidden by the last member
    var allowed = false
    for (ctype <- types) {
      if (matches(ctype)) {
        allowed = ctype.inclusive
      }
    }

    allowed
  }

  def paramValue(name: String): Option[String] = {
    val start = name + "="
    if (param.isDefined) {
      for (param <- param.get) {
        if (param.startsWith(start)) {
          return Some(param.substring(start.length))
        }
      }
    }
    None
  }

  def charset: Option[String] = {
    paramValue("charset")
  }

  def assertValid: MediaType = {
    if (assertValidName(mediaType) && assertValidName(mediaSubtype)) {
      return this
    }

    var ctype = mediaType
    if (mediaSubtype != "") {
      ctype += s"/${mediaSubtype}"
    }
    throw XProcException.xdUnrecognizedContentType(ctype, None)
  }

  private def assertValidName(literal: String): Boolean = {
    val name = literal.toLowerCase
    val length_ok = name.nonEmpty && name.length <= 127
    val start_ok = length_ok && name.substring(0,1).matches("^[a-z0-9]")
    val content_ok = start_ok && name.matches(s"[a-z0-9${Pattern.quote("!#$&-^_.+")}]+$$")
    content_ok
  }

  // https://alvinalexander.com/scala/how-to-define-equals-hashcode-methods-in-scala-object-equality
  private def canEqual(obj: Any): Boolean = obj.isInstanceOf[MediaType]

  override def equals(mtype: Any): Boolean =
    mtype match {
      case that: MediaType =>
        that.canEqual(this) &&
          this.mediaType == that.mediaType &&
          this.mediaSubtype == that.mediaSubtype
      case _ => false
    }

  override def hashCode(): Int = {
    val prime = 41
    var result = 1
    result = prime * result + mediaType.hashCode
    result = prime * result + mediaSubtype.hashCode
    result
  }

  override def toString: String = {
    var ctype = if (inclusive) {
      ""
    } else {
      "-"
    }
    ctype += mediaType + "/" + mediaSubtype
    if (suffix.isDefined) {
      ctype = ctype + "+" + suffix.get
    }
    if (param.isDefined) {
      for (param <- param.get) {
        ctype = ctype + "; " + param // Space after ; for compatibility with Morgana XProc results
      }
    }
    ctype
  }
}

File: xmlcalabash/src/main/kotlin/com/xmlcalabash/io/MediaType.kt
package com.xmlcalabash.io

import com.xmlcalabash.exceptions.XProcError
import com.xmlcalabash.util.MediaClassification
import net.sf.saxon.s9api.XdmAtomicValue
import java.nio.charset.Charset
import java.util.*
import java.util.regex.Pattern

class MediaType private constructor(val mediaType: String, val mediaSubtype: String,
                                    val suffix: String? = null, val inclusive: Boolean = true,
                                    val parameters: List<MediaTypeParameter> = listOf()) {
    companion object {
        val ANY = MediaType("*", "*")
        val GZIP = MediaType("application", "gzip")
        val HTML = MediaType("text", "html")
        val JPEG = MediaType("image", "jpeg")
        val JSON = MediaType("application", "json")
        val MULTIPART = MediaType("multipart", "*")
        val MULTIPART_MIXED = MediaType("multipart", "mixed")
        val OCTET_STREAM = MediaType("application", "octet-stream")
        val PDF = MediaType("application", "pdf")
        val PNG = MediaType("image", "png")
        val TEXT = MediaType("text", "plain")
        val XHTML = MediaType("application", "xhtml", "xml")
        val XML  = MediaType("application", "xml")
        val SVG = MediaType("image", "svg", "xml")
        val XQUERY = MediaType("application", "xquery")
        val XSLT = MediaType("application", "xslt", "xml")
        val YAML = MediaType("application", "vnd.yaml")
        val TOML = MediaType("application", "toml")
        val CSV = MediaType("text", "csv")
        val ZIP = MediaType("application", "zip")
        val JAR = MediaType("application", "java-archive")
        val TAR = MediaType("application", "x-tar")
        val AR = MediaType("application", "x-archive")
        val ARJ = MediaType("application", "x-arj")
        val CPIO = MediaType("application", "x-cpio")
        val SEVENZ = MediaType("application", "x-7z-compressed")
        val JAVA_PROPERTIES = MediaType("text", "x-java-properties")

        val JSONLD = MediaType("application", "ld", "json")
        val N3 = MediaType("text", "n3")
        val NQUADS = MediaType("application", "n-quads")
        val NTRIPLES = MediaType("application", "n-triples")
        val RDFJSON = MediaType("application", "rdf", "json")
        val RDFTHRIFT = MediaType("application", "rdf", "thrift")
        val RDFXML = MediaType("application", "rdf", "xml")
        val SPARQL = MediaType("application", "sparql-query")
        val SPARQL_RESULTS_XML = MediaType("application", "sparql-results", "xml")
        val SPARQL_RESULTS_JSON = MediaType("application", "sparql-results", "json")
        val TRIG = MediaType("application", "trig")
        val TRIX = MediaType("application", "trix", "xml")
        val TURTLE = MediaType("text", "turtle")

        val XML_OR_HTML = listOf<MediaType>(
            MediaType("application", "xml"),
            MediaType("text", "xml"),
            MediaType("*", "*", "xml")
        )

        val MATCH_XML = listOf<MediaType>(
            MediaType("application", "xml"),
            MediaType("text", "xml"),
            MediaType("*", "*", "xml"),
            MediaType("application", "xhtml", "xml", inclusive = false)
        )

        val MATCH_NOT_XML = listOf<MediaType>(
            MediaType("application", "xml", inclusive = false),
            MediaType("text", "xml", inclusive = false),
            MediaType("*", "*", "xml", inclusive = false)
        )

        val MATCH_HTML = listOf<MediaType>(
            MediaType("text", "html"),
            MediaType("application", "xhtml", "xml")
        )

        val MATCH_NOT_HTML = listOf<MediaType>(
            MediaType("text", "html", inclusive = false),
            MediaType("application", "xhtml", "xml", inclusive = false),
        )

        val MATCH_TEXT = listOf<MediaType>(
            MediaType("text", "*"),
            MediaType("application", "relax-ng-compact-syntax"),
            MediaType("application", "xquery"),
            MediaType("application", "javascript"),
            MediaType("application", "trig"),
            MediaType("application", "n-triples"),
            MediaType("application", "n-quads"),
            MediaType("application", "sparql-query"),
            MediaType("text", "html", inclusive = false),
            MediaType("text", "xml", inclusive = false),
            MediaType("text", "yaml", inclusive = false)
        )

        val MATCH_NOT_TEXT = listOf<MediaType>(
            MediaType("text", "*", inclusive = false),
            MediaType("application", "relax-ng-compact-syntax", inclusive = false),
            MediaType("application", "xquery", inclusive = false),
            MediaType("application", "javascript", inclusive = false),
        )

        val MATCH_JSON = listOf<MediaType>(
            MediaType("application", "json"),
            MediaType("*", "*", "json")
        )

        val MATCH_NOT_JSON = listOf<MediaType>(
            MediaType("application", "json", inclusive = false)
        )

        val MATCH_YAML = listOf<MediaType>(
            MediaType("application", "vnd.yaml"),
            MediaType("text", "yaml"),
            MediaType("application", "yaml"),
            MediaType("application", "x-yaml"),
            MediaType("application", "*", "yaml"),
        )

        val MATCH_TOML = listOf<MediaType>(
            MediaType("application", "toml")
        )

        val MATCH_CSV = listOf<MediaType>(
            MediaType("text", "csv")
        )

        val MATCH_ANY = listOf<MediaType>(
            MediaType("*", "*")
        )

        val MATCH_NOT_ANY = listOf<MediaType>(
            MediaType("*", "*", inclusive = false)
        )

        val classifications = listOf<Pair<MediaClassification, List<MediaType>>> (
            Pair(MediaClassification.XHTML, listOf(XHTML)),
            Pair(MediaClassification.HTML, MATCH_HTML),
            Pair(MediaClassification.XML, MATCH_XML),
            Pair(MediaClassification.JSON, MATCH_JSON),
            Pair(MediaClassification.YAML, MATCH_YAML),
            Pair(MediaClassification.TOML, MATCH_TOML),
            Pair(MediaClassification.TEXT, MATCH_TEXT),
        )

        fun parse(mtype: String?): MediaType? {
            if (mtype == null) {
                return null
            } else {
                return parse(mtype)
            }
        }

        fun parse(mtype: String, forceEncoding: String? = null): MediaType {
            var ctype = parseMatch(mtype, forceEncoding)
            if (!"[A-Za-z0-9][-A-Za-z0-9!#\$&0^_\\\\.+]*".toRegex().matches(ctype.mediaType) || ctype.mediaType.length > 127) {
                throw XProcError.xdInvalidContentType(mtype).exception()
            }
            if (!"[A-Za-z0-9][-A-Za-z0-9!#\$&0^_\\\\.+]*".toRegex().matches(ctype.mediaSubtype) || ctype.mediaSubtype.length > 127) {
                throw XProcError.xdInvalidContentType(mtype).exception()
            }
            return ctype
        }

        fun parseMatch(mtype: String, forceEncoding: String? = null): MediaType {
            // [-]type/subtype+ext; name1=val1; name2=val2
            var pos = mtype.indexOf("/")
            if (pos < 0) {
                throw XProcError.xdInvalidContentType(mtype).exception()
            }

            var mediaType = mtype.substring(0, pos).trim()
            var rest = mtype.substring(pos + 1).trim()
            var inclusive = true
            if (mediaType.startsWith("-")) {
                inclusive = false
                mediaType = mediaType.substring(1)
            }

            // This is a bit convoluted because of the way 'rest' is reused.
            // There was a bug and this was the easiest fix. #hack
            var params = ""
            pos = rest.indexOf(";")
            if (pos >= 0) {
                params = rest.substring(pos+1)
                rest = rest.substring(0, pos).trim()
            }

            var mediaSubtype = rest
            var suffix: String? = null

            if (mediaSubtype.contains("+")) {
                pos = mediaSubtype.indexOf("+")
                suffix = mediaSubtype.substring(pos+1).trim()
                mediaSubtype = mediaSubtype.substring(0, pos).trim()
            }

            if (forceEncoding == null && params == "") {
                return MediaType(mediaType, mediaSubtype, suffix, inclusive = inclusive)
            }

            rest = params
            pos = rest.indexOf(";")
            val plist = mutableListOf<MediaTypeParameter>()
            if (forceEncoding != null) {
                plist.add(MediaTypeParameter("charset=" + forceEncoding))
            }

            while (pos >= 0) {
                val param = rest.substring(0, pos).trim()
                rest = rest.substring(pos+1)
                if (param != "") {
                    if (forceEncoding == null || !param.startsWith("charset=")) {
                        plist.add(MediaTypeParameter(param))
                    }
                }
                pos = rest.indexOf(";")
            }

            if (rest.isNotBlank()) {
                rest = rest.trim()
                if (forceEncoding == null || !rest.startsWith("charset=")) {
                    plist.add(MediaTypeParameter(rest))
                }
            }

            return MediaType(mediaType, mediaSubtype, suffix, inclusive, plist.toList())
        }

        fun parseList(ctypes: String): List<MediaType> {
            val mlist = mutableListOf<MediaType>()
            for (ctype in ctypes.split("\\s+".toRegex())) {
                val excl = ctype[0] == '-'
                val type = if (excl) {
                    ctype.substring(1)
                } else {
                    ctype
                }

                when (type) {
                    "xml" -> if (excl) mlist.addAll(MATCH_NOT_XML) else mlist.addAll(MATCH_XML)
                    "html" -> if (excl) mlist.addAll(MATCH_NOT_HTML) else mlist.addAll(MATCH_HTML)
                    "text" -> if (excl) mlist.addAll(MATCH_NOT_TEXT) else mlist.addAll(MATCH_TEXT)
                    "json" -> if (excl) mlist.addAll(MATCH_NOT_JSON) else mlist.addAll(MATCH_JSON)
                    "any" -> if (excl) mlist.addAll(MATCH_NOT_ANY) else mlist.addAll(MATCH_ANY)
                    else -> {
                        if (ctype.contains("/")) {
                            mlist.add(parse(ctype))
                        } else {
                            throw XProcError.xsInvalidContentType(ctype).exception()
                        }
                    }
                }
            }
            return mlist
        }
    }

    fun discardParameters(): MediaType {
        if (parameters.isEmpty()) {
            return this
        }
        return MediaType(mediaType, mediaSubtype, suffix, inclusive)
    }

    fun discardParameters(paramNames: List<String>): MediaType {
        var discarded = false
        val newParams = mutableListOf<MediaTypeParameter>()
        for (param in parameters) {
            val name = param.name
            if (paramNames.contains(name)) {
                discarded = true
            } else {
                newParams.add(param)
            }
        }
        if (!discarded) {
            return this
        }
        return MediaType(mediaType, mediaSubtype, suffix, inclusive, newParams.toList())
    }

    fun addParam(name: String, value: String): MediaType {
        val newParams = mutableListOf<MediaTypeParameter>()
        for (param in parameters) {
            if (param.name != name) {
                newParams.add(param)
            }
        }
        newParams.add(MediaTypeParameter(name, value))
        return MediaType(mediaType, mediaSubtype, suffix, inclusive, newParams.toList())
    }

    fun matchingMediaType(mtypes: List<MediaType>): MediaType? {
        var match: MediaType? = null

        // Go through the whole list in case there are exclusions as well as inclusions
        for (mtype in mtypes) {
            if (matches(mtype)) {
                match = mtype
            }
        }

        return match
    }

    fun matches(mtype: MediaType): Boolean  {
        if (mtype.mediaType == "application" && mtype.mediaSubtype == "octet-stream") {
            return true
        }

        // application/xml should match */*
        // but text/plain shouldn't match */*+xml
/*
        // I wonder how many special cases there really are...can this be generalized somehow?
        if (testMatching(mtype, "xml", "xml")) {
            return true
        }
        if (testMatching(mtype, "json", "json")) {
            return true
        }
*/
        var mmatch = mediaType == mtype.mediaType || mtype.mediaType == "*"
        mmatch = mmatch && (mediaSubtype == mtype.mediaSubtype || mtype.mediaSubtype == "*")
        mmatch = mmatch && (mtype.suffix == null || suffix == mtype.suffix)
        return mmatch
    }

    private fun testMatching(mtype: MediaType, hassubtype: String, hassuffix: String): Boolean {
        if ((mediaType == "text" || mediaType == "application") && (mediaSubtype == hassubtype || suffix == hassuffix)) {
            if ((mtype.mediaType == "text" || mtype.mediaType == "application")
                && (mtype.mediaSubtype == hassubtype || mtype.suffix == hassuffix)) {
                return true
            }
        }
        return false
    }

    fun classification(): MediaClassification {
        for ((type, typeList) in classifications) {
            val mtype = matchingMediaType(typeList)
            if (mtype != null && mtype.inclusive) {
                return type
            }
        }
        return MediaClassification.BINARY
    }

    fun allowed(types: List<MediaType>): Boolean {
        // This media type is allowed if it's allowed by at least one member of types
        // and is not forbidden by the last member
        var allowed = false
        for (ctype in types) {
            if (matches(ctype)) {
                allowed = ctype.inclusive
            }
        }

        return allowed
    }

    fun paramValue(name: String): String? {
        for (param in parameters) {
            if (param.name == name) {
                val value = param.value
                if ((value.startsWith("\"") && value.endsWith("\""))
                    || (value.startsWith("'") && value.endsWith("'"))) {
                    return value.substring(1, value.length - 1)
                }
                return value
            }
        }
        return null
    }

    fun charset(): Charset? {
        val name = paramValue("charset")
        if (name == null) {
            return null
        }

        if (!Charset.isSupported(name)) {
            throw XProcError.xdUnsupportedDocumentCharset(name).exception()
        }

        return Charset.forName(name)
    }

    fun assertValid(): MediaType {
        if (assertValidName(mediaType) && assertValidName(mediaSubtype)) {
            return this
        }
        throw RuntimeException("Unrecognized content type: ${mediaType}/${mediaSubtype}")
    }

    private fun assertValidName(literal: String): Boolean {
        val name = literal.lowercase(Locale.getDefault())
        val lengthOk = name.isNotEmpty() && name.length <= 127
        val startOk = lengthOk && name.substring(0,1).matches("^[a-z0-9]".toRegex())
        val contentOk = startOk && name.matches("[a-z0-9${Pattern.quote("!#$&-^_.+")}]+$$".toRegex())
        return contentOk
    }

    override fun equals(other: Any?): Boolean {
        return when (other) {
            is MediaType -> {
                if (mediaType != other.mediaType || mediaSubtype != other.mediaSubtype) {
                    return false
                }

                return suffix == other.suffix
            }
            else -> false
        }
    }

    override fun hashCode(): Int {
        return (41 * mediaType.hashCode()) + mediaSubtype.hashCode()
    }

    fun toAtomicValue(): XdmAtomicValue {
        return XdmAtomicValue(toString())
    }

    fun toStringWithoutParameters(): String {
        if (suffix == null) {
            return "${mediaType}/${mediaSubtype}"
        }
        return "${mediaType}/${mediaSubtype}+${suffix}"
    }

    override fun toString(): String {
        val sb = StringBuilder()
        if (!inclusive) {
            sb.append("-")
        }
        sb.append(toStringWithoutParameters())

        for (param in parameters) {
            // Space after ; for compatibility with Morgana XProc results
            sb.append("; ").append(param)
        }

        return sb.toString()
    }
}
